File: src/packages/arcane/lib/deficit-governance.mjs
// Deterministic deficit & outcome governance.  This module only accepts
// explicit structured classification fields; it never infers debt from prose.
import { decision } from './errors.mjs';

const DEBT_SAFE_CLASSES = new Set(['OPTIONAL', 'QUALITY', 'PERFORMANCE', 'USABILITY']);
const PROTECTED_CLASSES = new Set(['CORRECTNESS', 'SAFETY']);
const PASS = 'PASS';
const nonEmpty = (value) => typeof value === 'string' && value.length > 0;
const freeze = (value) => Object.freeze(value);
const deny = (code, message, detail = {}) => decision({ allowed: false, code, message, detail });

function normalizeAcceptance(item, index) {
  const id = item?.acceptanceId ?? item?.acceptance_id ?? item?.id;
  const obligationClass = item?.obligationClass ?? item?.obligation_class;
  const result = item?.result;
  if (!nonEmpty(id) || !nonEmpty(obligationClass) || !nonEmpty(result)) {
    return null;
  }
  return freeze({ id, obligationClass, result, canonicalDeficitId: `deficit:${id}` });
}

// Classify every non-passing acceptance item.  Only explicit optional/quality
// classes may become debt; correctness & safety always retain their gate.
export function classifyAcceptanceDeficits({ acceptanceItems = [] } = {}) {
  if (!Array.isArray(acceptanceItems)) return deny('ARC_SCHEMA_INVALID', 'acceptanceItems must be an array');
  const normalized = acceptanceItems.map(normalizeAcceptance);
  if (normalized.some((item) => item === null)) return deny('ARC_SCHEMA_INVALID', 'every acceptance item requires id, obligationClass, and result');
  const deficits = normalized.filter((item) => item.result !== PASS).map((item) => freeze({
    ...item,
    debtEligible: DEBT_SAFE_CLASSES.has(item.obligationClass),
    protected: PROTECTED_CLASSES.has(item.obligationClass),
    disposition: DEBT_SAFE_CLASSES.has(item.obligationClass) ? 'DEBT' : 'BLOCKING',
  }));
  const blocking = deficits.filter((item) => item.disposition === 'BLOCKING');
  const debt = deficits.filter((item) => item.disposition === 'DEBT');
  const claimCeiling = blocking.length ? 'CANDIDATE' : debt.length ? 'COMPLETE_WITH_DEBT' : 'COMPLETE';
  return decision({
    allowed: true,
    message: 'acceptance deficits classified',
    detail: freeze({ deficits: freeze(deficits), claimCeiling, outcomeState: claimCeiling === 'COMPLETE' ? 'COMPLETE' : claimCeiling === 'COMPLETE_WITH_DEBT' ? 'COMPLETE_WITH_DEBT' : 'CANDIDATE' }),
  });
}

// Debt is a conversion, not a label.  This guard rejects any protected or
// required class before a caller can expose a downgraded completion claim.
export function convertDeficitsToDebt(input = {}) {
  const classified = classifyAcceptanceDeficits(input);
  if (!classified.allowed) return classified;
  const protectedDeficits = classified.detail.deficits.filter((deficit) => deficit.protected);
  if (protectedDeficits.length) {
    return deny('ARC_CLAIM_PREREQUISITE_UNMET', 'correctness or safety deficit cannot convert to debt', freeze({
      outcomeState: 'CANDIDATE', claimCeiling: 'CANDIDATE',
      protectedDeficitIds: freeze(protectedDeficits.map((deficit) => deficit.canonicalDeficitId)),
    }));
  }
  const blocking = classified.detail.deficits.filter((deficit) => deficit.disposition === 'BLOCKING');
  if (blocking.length) return deny('ARC_CLAIM_PREREQUISITE_UNMET', 'required acceptance deficit cannot convert to debt', freeze({ outcomeState: 'CANDIDATE', claimCeiling: 'CANDIDATE', blockingDeficitIds: freeze(blocking.map((deficit) => deficit.canonicalDeficitId)) }));
  return decision({ allowed: true, message: 'eligible acceptance deficits converted to canonical debt', detail: freeze({ outcomeState: classified.detail.outcomeState, claimCeiling: classified.detail.claimCeiling, debts: freeze(classified.detail.deficits) }) });
}

// A consumer may proceed with degraded behavior only by acknowledging a
// producer's canonical debt identity.  Locally copied debt objects are not an
// acknowledgement and cannot cross a dispatch boundary.
export function acknowledgeDownstreamDeficits({ canonicalDeficits = [], acknowledgements = [] } = {}) {
  if (!Array.isArray(canonicalDeficits) || !Array.isArray(acknowledgements)) return deny('ARC_SCHEMA_INVALID', 'canonicalDeficits and acknowledgements must be arrays');
  const canonicalIds = new Set(canonicalDeficits.map((deficit) => deficit?.canonicalDeficitId).filter(nonEmpty));
  if (canonicalIds.size !== canonicalDeficits.length) return deny('ARC_SCHEMA_INVALID', 'each canonical deficit requires canonicalDeficitId');
  const acknowledged = new Set();
  for (const acknowledgement of acknowledgements) {
    if (!nonEmpty(acknowledgement?.canonicalDeficitId) || acknowledgement?.canonical !== true || !nonEmpty(acknowledgement?.consumerId)) return deny('ARC_SCHEMA_INVALID', 'acknowledgement requires canonical identity, canonical marker, and consumerId');
    if (!canonicalIds.has(acknowledgement.canonicalDeficitId)) return deny('ARC_BINDING_MISMATCH', 'downstream acknowledgement does not reference canonical debt', { canonicalDeficitId: acknowledgement.canonicalDeficitId });
    acknowledged.add(acknowledgement.canonicalDeficitId);
  }
  const missing = [...canonicalIds].filter((id) => !acknowledged.has(id));
  if (missing.length) return deny('ARC_CLAIM_PREREQUISITE_UNMET', 'dispatch rejected until canonical debt is acknowledged', freeze({ dispatch: 'REJECTED', missingCanonicalDeficitIds: freeze(missing) }));
  return decision({ allowed: true, message: 'downstream acknowledged canonical deficits', detail: freeze({ dispatch: 'ALLOWED', acknowledgedCanonicalDeficitIds: freeze([...acknowledged].sort()) }) });
}

// Internal tests are supporting evidence only.  A complete outcome requires a
// structured, authenticated observation at its declared acceptance surface.
export function evaluateOutcomeClosure({ acceptanceSurface } = {}) {
  const observed = acceptanceSurface?.observed === true;
  const authenticated = acceptanceSurface?.authenticated === true;
  const stateIdentity = acceptanceSurface?.integratedState;
  if (!observed || !authenticated || stateIdentity === null || stateIdentity === undefined) {
    return deny('ARC_EVIDENCE_INSUFFICIENT', 'outcome remains CANDIDATE until acceptance surface is observed', freeze({ outcomeState: 'CANDIDATE', claimCeiling: 'CANDIDATE', acceptanceSurface: observed ? 'UNAUTHENTICATED' : 'UNOBSERVED' }));
  }
  return decision({ allowed: true, message: 'outcome acceptance surface observed', detail: freeze({ outcomeState: 'COMPLETE', claimCeiling: 'COMPLETE', integratedState: stateIdentity }) });
}

File: engine/crates/legion-arcane/src/deficit_governance.rs
use crate::decision::{decision, deny};
use serde_json::{json, Value};
use std::collections::HashSet;

const PASS: &str = "PASS";

fn non_empty(value: &Value) -> bool {
    value.as_str().is_some_and(|text| !text.is_empty())
}

fn normalize_acceptance(item: &Value) -> Option<Value> {
    let id = item
        .get("acceptanceId")
        .or_else(|| item.get("acceptance_id"))
        .or_else(|| item.get("id"))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty());
    let obligation_class = item
        .get("obligationClass")
        .or_else(|| item.get("obligation_class"))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty());
    let result = item
        .get("result")
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty());
    if id.is_none() || obligation_class.is_none() || result.is_none() {
        return None;
    }
    Some(json!({
        "id": id,
        "obligationClass": obligation_class,
        "result": result,
        "canonicalDeficitId": format!("deficit:{}", id.unwrap()),
    }))
}

pub fn classify_acceptance_deficits(input: &Value) -> Value {
    let acceptance_items = input.get("acceptanceItems").and_then(Value::as_array);
    if acceptance_items.is_none() {
        return deny(
            "ARC_SCHEMA_INVALID",
            "acceptanceItems must be an array",
            json!({}),
        );
    }
    let normalized = acceptance_items
        .unwrap()
        .iter()
        .map(normalize_acceptance)
        .collect::<Vec<_>>();
    if normalized.iter().any(|item| item.is_none()) {
        return deny(
            "ARC_SCHEMA_INVALID",
            "every acceptance item requires id, obligationClass, and result",
            json!({}),
        );
    }
    let deficits = normalized
        .into_iter()
        .flatten()
        .filter(|item| item.get("result").and_then(Value::as_str) != Some(PASS))
        .map(|item| {
            let obligation = item["obligationClass"].as_str().unwrap_or_default();
            let debt_eligible = matches!(obligation, "OPTIONAL" | "QUALITY" | "PERFORMANCE" | "USABILITY");
            let protected = matches!(obligation, "CORRECTNESS" | "SAFETY");
            let disposition = if debt_eligible { "DEBT" } else { "BLOCKING" };
            json!({
                "id": item["id"],
                "obligationClass": item["obligationClass"],
                "result": item["result"],
                "canonicalDeficitId": item["canonicalDeficitId"],
                "debtEligible": debt_eligible,
                "protected": protected,
                "disposition": disposition,
            })
        })
        .collect::<Vec<_>>();
    let blocking = deficits
        .iter()
        .filter(|item| item.get("disposition") == Some(&json!("BLOCKING")))
        .cloned()
        .collect::<Vec<_>>();
    let debt = deficits
        .iter()
        .filter(|item| item.get("disposition") == Some(&json!("DEBT")))
        .cloned()
        .collect::<Vec<_>>();
    let claim_ceiling = if !blocking.is_empty() {
        "CANDIDATE"
    } else if !debt.is_empty() {
        "COMPLETE_WITH_DEBT"
    } else {
        "COMPLETE"
    };
    let outcome_state = claim_ceiling;
    decision(
        true,
        None,
        Some("acceptance deficits classified"),
        json!({
            "deficits": deficits,
            "claimCeiling": claim_ceiling,
            "outcomeState": outcome_state,
        }),
    )
}

pub fn convert_deficits_to_debt(input: &Value) -> Value {
    let classified = classify_acceptance_deficits(input);
    if classified.get("allowed").and_then(Value::as_bool) != Some(true) {
        return classified;
    }
    let detail = classified.get("detail").cloned().unwrap_or(Value::Null);
    let deficits = detail.get("deficits").and_then(Value::as_array).cloned().unwrap_or_default();
    let protected = deficits
        .iter()
        .filter(|deficit| deficit.get("protected") == Some(&json!(true)))
        .collect::<Vec<_>>();
    if !protected.is_empty() {
        return deny(
            "ARC_CLAIM_PREREQUISITE_UNMET",
            "correctness or safety deficit cannot convert to debt",
            json!({
                "outcomeState": "CANDIDATE",
                "claimCeiling": "CANDIDATE",
                "protectedDeficitIds": protected
                    .iter()
                    .filter_map(|deficit| deficit.get("canonicalDeficitId").cloned())
                    .collect::<Vec<_>>(),
            }),
        );
    }
    let blocking = deficits
        .iter()
        .filter(|deficit| deficit.get("disposition") == Some(&json!("BLOCKING")))
        .collect::<Vec<_>>();
    if !blocking.is_empty() {
        return deny(
            "ARC_CLAIM_PREREQUISITE_UNMET",
            "required acceptance deficit cannot convert to debt",
            json!({
                "outcomeState": "CANDIDATE",
                "claimCeiling": "CANDIDATE",
                "blockingDeficitIds": blocking
                    .iter()
                    .filter_map(|deficit| deficit.get("canonicalDeficitId").cloned())
                    .collect::<Vec<_>>(),
            }),
        );
    }
    decision(
        true,
        None,
        Some("eligible acceptance deficits converted to canonical debt"),
        json!({
            "outcomeState": detail.get("outcomeState").cloned().unwrap_or(Value::Null),
            "claimCeiling": detail.get("claimCeiling").cloned().unwrap_or(Value::Null),
            "debts": deficits,
        }),
    )
}

pub fn acknowledge_downstream_deficits(input: &Value) -> Value {
    let canonical_deficits = input.get("canonicalDeficits").and_then(Value::as_array);
    let acknowledgements = input.get("acknowledgements").and_then(Value::as_array);
    if canonical_deficits.is_none() || acknowledgements.is_none() {
        return deny(
            "ARC_SCHEMA_INVALID",
            "canonicalDeficits and acknowledgements must be arrays",
            json!({}),
        );
    }
    let canonical_ids = canonical_deficits
        .unwrap()
        .iter()
        .filter_map(|deficit| deficit.get("canonicalDeficitId").and_then(Value::as_str))
        .filter(|value| !value.is_empty())
        .collect::<Vec<_>>();
    if canonical_ids.len() != canonical_deficits.unwrap().len() {
        return deny(
            "ARC_SCHEMA_INVALID",
            "each canonical deficit requires canonicalDeficitId",
            json!({}),
        );
    }
    let canonical_set = canonical_ids.iter().collect::<HashSet<_>>();
    let mut acknowledged = HashSet::<String>::new();
    for acknowledgement in acknowledgements.unwrap() {
        if !non_empty(&acknowledgement["canonicalDeficitId"])
            || acknowledgement.get("canonical") != Some(&json!(true))
            || !non_empty(&acknowledgement["consumerId"])
        {
            return deny(
                "ARC_SCHEMA_INVALID",
                "acknowledgement requires canonical identity, canonical marker, and consumerId",
                json!({}),
            );
        }
        let id = acknowledgement["canonicalDeficitId"]
            .as_str()
            .unwrap_or_default()
            .to_owned();
        if !canonical_set.contains(&id.as_str()) {
            return deny(
                "ARC_BINDING_MISMATCH",
                "downstream acknowledgement does not reference canonical debt",
                json!({ "canonicalDeficitId": id }),
            );
        }
        acknowledged.insert(id);
    }
    let missing = canonical_set
        .iter()
        .filter(|id| !acknowledged.contains(&id.to_string()))
        .cloned()
        .collect::<Vec<_>>();
    if !missing.is_empty() {
        return deny(
            "ARC_CLAIM_PREREQUISITE_UNMET",
            "dispatch rejected until canonical debt is acknowledged",
            json!({
                "dispatch": "REJECTED",
                "missingCanonicalDeficitIds": missing,
            }),
        );
    }
    let mut ids = acknowledged.into_iter().collect::<Vec<_>>();
    ids.sort();
    decision(
        true,
        None,
        Some("downstream acknowledged canonical deficits"),
        json!({
            "dispatch": "ALLOWED",
            "acknowledgedCanonicalDeficitIds": ids,
        }),
    )
}

pub fn evaluate_outcome_closure(input: &Value) -> Value {
    let acceptance_surface = input.get("acceptanceSurface").unwrap_or(&Value::Null);
    let observed = acceptance_surface.get("observed") == Some(&json!(true));
    let authenticated = acceptance_surface.get("authenticated") == Some(&json!(true));
    let state_identity = acceptance_surface.get("integratedState");
    if !observed
        || !authenticated
        || state_identity.is_none()
        || state_identity == Some(&Value::Null)
    {
        return deny(
            "ARC_EVIDENCE_INSUFFICIENT",
            "outcome remains CANDIDATE until acceptance surface is observed",
            json!({
                "outcomeState": "CANDIDATE",
                "claimCeiling": "CANDIDATE",
                "acceptanceSurface": if observed { "UNAUTHENTICATED" } else { "UNOBSERVED" },
            }),
        );
    }
    decision(
        true,
        None,
        Some("outcome acceptance surface observed"),
        json!({
            "outcomeState": "COMPLETE",
            "claimCeiling": "COMPLETE",
            "integratedState": state_identity,
        }),
    )
}
